community/kernel/src/main/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanReader.java

```diff
@@ -0,0 +1,165 @@
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.index.labelscan;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+
+import org.neo4j.collection.primitive.PrimitiveLongIterator;
+import org.neo4j.cursor.RawCursor;
+import org.neo4j.index.Hit;
+import org.neo4j.index.Index;
+import org.neo4j.index.gbptree.GBPTree;
+import org.neo4j.storageengine.api.schema.LabelScanReader;
+
+/**
+ * {@link LabelScanReader} for reading data from {@link NativeLabelScanStore}.
+ * Each {@link PrimitiveLongIterator} returned from each of the methods is backed by {@link RawCursor}
+ * directly from {@link GBPTree#seek(Object, Object)}.
+ * <p>
+ * The returned {@link PrimitiveLongIterator} aren't closable so the cursors retrieved are managed
+ * inside of this reader and closed between each new query and on {@link #close()}.
+ */
+class NativeLabelScanReader implements LabelScanReader
+{
+    /**
+     * {@link Index} which is queried when calling the methods below.
+     */
+    private final Index<LabelScanKey,LabelScanValue> index;
+
+    /**
+     * Size of each nodeId range, i.e. how many potential node ids each {@link LabelScanValue} contains.
+     */
+    private final int rangeSize;
+
+    /**
+     * Currently open {@link RawCursor} from query methods below. Open cursors are closed when calling
+     * new query methods or when {@link #close() closing} this reader.
+     */
+    private final Queue<RawCursor<Hit<LabelScanKey,LabelScanValue>,IOException>> openCursors;
+
+    NativeLabelScanReader( Index<LabelScanKey,LabelScanValue> index, int rangeSize )
+    {
+        this.index = index;
+        this.rangeSize = rangeSize;
+        this.openCursors = new LinkedList<>();
+    }
+
+    /**
+     * Closes all currently open {@link RawCursor cursors} from last query method call.
+     */
+    @Override
+    public void close()
+    {
+        try
+        {
+            ensureOpenCursorsClosed();
+        }
+        catch ( IOException e )
+        {
+            throw new RuntimeException( e );
+        }
+    }
+
+    @Override
+    public PrimitiveLongIterator nodesWithLabel( int labelId )
+    {
+        RawCursor<Hit<LabelScanKey,LabelScanValue>,IOException> cursor;
+        try
+        {
+            ensureOpenCursorsClosed();
```

Comment:
That would be a bigger change to the neo4j integration. The current behaviour is what's expected of the `LabelScanReader` afaict. Let's revisit this behaviour in another context if need be, shall we?